android-core/src/main/kotlin/com/mparticle/internal/UserStorage.kt

```diff
@@ -0,0 +1,604 @@
+        return mPreferences.contains(PREVIOUS_SESSION_ID)
+    }
+
+    fun getPreviousSessionStart(defaultValue: Long): Long {
+        return mPreferences.getLong(PREVIOUS_SESSION_START, defaultValue)
+    }
+
+    fun setPreviousSessionStart(previousSessionStart: Long) {
+        mPreferences.edit().putLong(PREVIOUS_SESSION_START, previousSessionStart).apply()
+    }
+
+    private fun hasPreviousSessionStart(): Boolean {
+        return mPreferences.contains(PREVIOUS_SESSION_START)
+    }
+
+    var ltv: String?
+        get() = mPreferences.getString(LTV, "0")
+        set(ltv) {
+            mPreferences.edit().putString(LTV, ltv).apply()
+        }
+
+    private fun hasLtv(): Boolean {
+        return mPreferences.contains(LTV)
+    }
+
+    fun getTotalRuns(defaultValue: Int): Int {
+        return mPreferences.getInt(TOTAL_RUNS, defaultValue)
+    }
+
+    fun setTotalRuns(totalRuns: Int) {
+        mPreferences.edit().putInt(TOTAL_RUNS, totalRuns).apply()
+    }
+
+    private fun hasTotalRuns(): Boolean {
+        return mPreferences.contains(TOTAL_RUNS)
+    }
+
+    var cookies: String?
+        get() = mPreferences.getString(COOKIES, "")
+        set(cookies) {
+            mPreferences.edit().putString(COOKIES, cookies).apply()
+        }
+
+    private fun hasCookies(): Boolean {
+        return mPreferences.contains(COOKIES)
+    }
+
+    var launchesSinceUpgrade: Int
+        get() = mPreferences.getInt(TOTAL_SINCE_UPGRADE, 0)
+        set(launchesSinceUpgrade) {
+            mPreferences.edit().putInt(TOTAL_SINCE_UPGRADE, launchesSinceUpgrade).apply()
+        }
+
+    private fun hasLaunchesSinceUpgrade(): Boolean {
+        return mPreferences.contains(TOTAL_SINCE_UPGRADE)
+    }
+
+    var userIdentities: String?
+        get() = mPreferences.getString(USER_IDENTITIES, "")
+        set(userIdentities) {
+            mPreferences.edit().putString(USER_IDENTITIES, userIdentities).apply()
+        }
+
+    var serializedConsentState: String?
+        get() = mPreferences.getString(CONSENT_STATE, null)
+        set(consentState) {
+            mPreferences.edit().putString(CONSENT_STATE, consentState).apply()
+        }
+
+    private fun hasConsent(): Boolean {
+        return mPreferences.contains(CONSENT_STATE)
+    }
+
+    val isLoggedIn: Boolean
+        get() = mPreferences.getBoolean(KNOWN_USER, false)
+
+    var firstSeenTime: Long?
+        get() {
+            if (!mPreferences.contains(FIRST_SEEN_TIME)) {
+                mPreferences.edit().putLong(
+                    FIRST_SEEN_TIME, messageManagerSharedPreferences.getLong(
+                        Constants.PrefKeys.INSTALL_TIME, defaultSeenTime
+                    )
+                ).apply()
+            }
+            return mPreferences.getLong(FIRST_SEEN_TIME, defaultSeenTime)
+        }
+        set(time) {
+            if (!mPreferences.contains(FIRST_SEEN_TIME)) {
+                mPreferences.edit().putLong(FIRST_SEEN_TIME, time!!).apply()
```

Comment:
Looks like we're force unwrapping time without checking for null, and we explicitly allow null since this is marked as `Long?`. So this will throw a null pointer exception whenever `time` is null...

We should never be using `!!` anywhere in production code, only in tests.